python/gps/algorithm/algorithm_mdgps.py

```diff
@@ -0,0 +1,334 @@
+""" This file defines the MD-based GPS algorithm. """
+import copy
+import logging
+
+import numpy as np
+import scipy as sp
+
+from gps.algorithm.algorithm import Algorithm
+from gps.algorithm.algorithm_utils import PolicyInfo, gauss_fit_joint_prior
+from gps.algorithm.config import ALG_MDGPS
+from gps.sample.sample_list import SampleList
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AlgorithmMDGPS(Algorithm):
+    """
+    Sample-based joint policy learning and trajectory optimization with
+    (approximate) mirror descent guided policy search algorithm.
+    """
+    def __init__(self, hyperparams):
+        config = copy.deepcopy(ALG_MDGPS)
+        config.update(hyperparams)
+        Algorithm.__init__(self, config)
+
+        for m in range(self.M):
+            self.cur[m].pol_info = PolicyInfo(self._hyperparams)
+            policy_prior = self._hyperparams['policy_prior']
+            self.cur[m].pol_info.policy_prior = \
+                    policy_prior['type'](policy_prior)
+
+        self.policy_opt = self._hyperparams['policy_opt']['type'](
+            self._hyperparams['policy_opt'], self.dO, self.dU
+        )
+
+    def iteration(self, sample_lists):
+        """
+        Run iteration of BADMM-based guided policy search.
+
+        Args:
+            sample_lists: List of SampleList objects for each condition.
+        """
+        # Store the samples.
+        for m in range(self.M):
+            self.cur[m].sample_list = sample_lists[m]
+
+        self._update_dynamics()  # Update dynamics model using all sample.
+        self._update_policy_samples()  # Choose samples to use with the policy.
+
+        # On the first iteration we need to make sure that the policy somewhat
+        # matches the init controller. Otherwise the LQR backpass starts with
+        # a bad linearization, and things don't work out well.
+        if self.iteration_count == 0:
+            self.new_traj_distr = [
+                self.cur[cond].traj_distr for cond in range(self.M)
+            ]
+            self._update_policy(self.iteration_count, 1) # HACK: need to set inner_itr=1
```

Comment:
It's a bit of a hack because it's not really inner_itr 1, you just have to say that so that the mean and variance will be set correctly. Agreed that adding a normalize flag is the better option, and I'll make that change. In the future we might be better off just using batch norm, especially since the conditions will be changing randomly, but let's put that off for now.